File: Second Year/COS212/Assignments/07_01_Assignment 4/Assignment 1_JAVA_FILES/OrderedContainer.java
abstract class OrderedContainer<T extends Comparable<T>> {

    OrderedContainer(LinearStructure<T> c) {
        this.dataStructure = c.clone();
    }

    OrderedContainer(OrderedContainer<T> other) {
        this.dataStructure = other.dataStructure.clone();
    }

    @Override
    public OrderedContainer<T> clone() {
        OrderedContainer<T> temp = null;
        temp.dataStructure = this.dataStructure.clone();
        return temp;
    }

    public abstract T remove();

    public abstract T next() throws RemoveException;

    public abstract void insert(T el);

    public boolean isEmpty() {
        return this.dataStructure.isEmpty();
    }

    public LinearStructure<T> getImplementation() {
        return this.dataStructure;
    }

    protected LinearStructure<T> dataStructure;

}

File: First Year/COS110/Assignments/Assignment 4/orderedContainer.cpp
#include "orderedContainer.h"

template <class T>
OrderedContainer<T>::OrderedContainer(LinearStructure<T> *c) {
    dataStructure = c->clone();
}

template <class T>
OrderedContainer<T>::OrderedContainer(const OrderedContainer<T> &other) {
    OrderedContainer<T> * newContainer;
    newContainer->dataStructure = other.dataStructure->clone();
}

template <class T>
OrderedContainer<T> & OrderedContainer<T>::operator=(const OrderedContainer<T> &other) {
    OrderedContainer<T> * newContainer;
    newContainer->dataStructure = other.dataStructure->clone();
    return *newContainer;
}

template <class T>
OrderedContainer<T>::~OrderedContainer<T>() {
    //delete dataStructure;
}

template <class T>
bool OrderedContainer<T>::isEmpty() {
    /*if (!dataStructure->isEmpty()) {
        return true;
    }
    return false;*/
    return dataStructure->isEmpty();
}

template <class T>
LinearStructure<T> * OrderedContainer<T>::getImplementation() {
    return this->dataStructure;
}
